components/api/src/main/java/com/hotels/styx/api/messages/FullHttpRequest.java

```diff
@@ -0,0 +1,694 @@
+import static io.netty.handler.codec.http.HttpMethod.PUT;
+import static io.netty.handler.codec.http.HttpMethod.TRACE;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+import static java.lang.Integer.parseInt;
+import static java.net.InetSocketAddress.createUnresolved;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+
+/**
+ * HTTP request with a fully aggregated/decoded body.
+ *
+ * @param <T> content type
+ */
+public class FullHttpRequest<T> implements FullHttpMessage<T> {
+    private final Object id;
+    private final InetSocketAddress clientAddress;
+    private final HttpVersion version;
+    private final HttpMethod method;
+    private final Url url;
+    private final HttpHeaders headers;
+    private final boolean secure;
+    private final T body;
+    private final List<HttpCookie> cookies;
+
+    FullHttpRequest(Builder<T> builder) {
+        this.id = builder.id();
+        this.clientAddress = builder.clientAddress();
+        this.version = builder.version();
+        this.method = builder.method();
+        this.url = builder.url();
+        this.secure = builder.secure();
+        this.headers = builder.headers();
+        this.body = builder.body();
+        this.cookies = builder.cookies();
+    }
+
+    /**
+     * Creates a request with the GET method.
+     *
+     * @param uri URI
+     * @return {@code this}
+     */
+    public static <T> Builder<T> get(String uri) {
+        return new Builder<>(GET, uri);
+    }
+
+    /**
+     * Creates a request with the HEAD method.
+     *
+     * @param uri URI
+     * @return {@code this}
+     */
+    public static <T> Builder<T> head(String uri) {
+        return new Builder<>(HEAD, uri);
+    }
+
+    /**
+     * Creates a request with the POST method.
+     *
+     * @param uri URI
+     * @return {@code this}
+     */
+    public static <T> Builder<T> post(String uri) {
+        return new Builder<>(POST, uri);
+    }
+
+    /**
+     * Creates a request with the DELETE method.
+     *
+     * @param uri URI
+     * @return {@code this}
+     */
+    public static <T> Builder<T> delete(String uri) {
+        return new Builder<>(DELETE, uri);
+    }
+
+    /**
+     * Creates a request with the PUT method.
+     *
+     * @param uri URI
+     * @return {@code this}
+     */
+    public static <T> Builder<T> put(String uri) {
+        return new Builder<>(PUT, uri);
+    }
+
+    /**
+     * Creates a request with the PATCH method.
+     *
+     * @param uri URI
+     * @return {@code this}
+     */
+    public static <T> Builder<T> patch(String uri) {
+        return new Builder<>(PATCH, uri);
+    }
+
+    /**
+     * Creates a request with the GET method.
+     *
+     * @param uri URI
+     * @param body body
+     * @param <T> body type
+     * @return {@code this}
+     */
+    public static <T> Builder<T> get(String uri, T body) {
```

Comment:
The GET and HEAD (below) methods are not intended for pushing content up to server. Therefore we should remove this overloaded variant and the another for `head(String, T)`.